# Analysis of next-generation sequencing data using **Cytoscape**



<img src="images/Cytoscape-Logo.png" style="position:absolute;top:40px;right:10px;" width="200" />

<img src="images/logo-sm.png" style="position:absolute;top:40px;center:10px;" width="200" />

## Workshop Information
- June 12th 2024 - University of Sheffield
- Bartolome House, Seminar Room EG03

## Schedule (Provisonal)

### Thursday 12th June, 2024

- 09:30 - 11:00 Introduction to Cytoscape and it's App
- 11:15 - 12:30 Functional Enrichment Analysis
- 12:30 - 13:15 LUNCH
- 13:15 - 15:00 Enrichment using ClueGO and CluePedia
- 15:00 - 16:30 Identifying signiifcant pathway modules and differential expression genes



You will need to attend all sessions to complete the workshop


***Some basic Network Biology knowledge is assumed (and is essential). Without it, you will struggle on this course. If you are not familiar with the key terms. We strongly encourage you to work through an introductory Cytoscape course before attempting these materials. We recommend reviewing Parts 1 and 2 of our Cytoscape introductory course before deciding if you can attend***

- [Cytoscape crash course Part 1](https://a1aks.github.io/Cytoscape_Course/Intro_NetworkBio.nb.html)
- [Cytoscape crash course Part 2](https://a1aks.github.io/Cytoscape_Course/session1a.nb.html) 

## Registration Link

You can sign-up to attend either in-person or online using this link

- [https://onlineshop.shef.ac.uk/conferences-and-events/faculty-of-medicine-dentistry-and-health/medical-school](https://onlineshop.shef.ac.uk/conferences-and-events/faculty-of-medicine-dentistry-and-health/medical-school)

If you are registering to attend as a University of Sheffield staff or student, you will be prompted for a password. This should have been sent to you in an announcment email.



## Materials

This course comprises of three sessions. 
- [Session 1:A - Introduction to Cytoscape and it's App - Part 1](session1a.Rmd)
- [Part 1:A - HTML](session1a.nb.html)
- [Session 1:B - Introduction to Cytoscape and it's App - Part 2](session1b.Rmd)
- [Part 1:B - HTML](session1b.nb.html)
- [Session 2:A - Basic network analysis using RNA-Seq data](session2a.Rmd)
- [Part 2:A - HTML](session2a.nb.html)
- [Session 2:B - Functional Enrichment Analysis](session2b.Rmd)
- [Part 2:B - HTML](session2b.nb.html)
- [Session 3 - Enrichment using ClueGO and CluePedia.](session3.Rmd)
- [Part 3 - HTML](session3.nb.html)



## Overview

The workshop is designed to give you an introduction to using Cytoscape, which is a network-based software application to perform bioinformatics data and pathway enrichment analysis. The particular use-case in this course is that of RNA-seq data, but many of the concepts are used for other types of **omics**-analysis. There are several reasons why we might want to use **Cytoscpae** in **Bioinformatics**:-
* You can be a naive or non-programmer for analysing your biological datasets generated by “high-throughput sequencing” through dedicated Cytoscape packages.
* The **packages/Apps** within Cytoscape provide a user complete support for different types of “-Omics” datasets.
 *	e.g. many NGS tools are developed for a Unix environment.
 * Some tasks require more memory (RAM) and processor speed than we have on our own machine
 * Some analyses take too long, and whilst they are running everything else on our machine will be slow.
 * Eventually you will want to use “High-performance computing” (HPC) - but this requires some background knowledge before you can use it.
* Hence, **Cytoscape** provides many advantages for those users who wish to have their biological data analysed in a more user-freindly manner but also visualizing results simultaneously. 

## Who should attend this course?

Researchers in life sciences who want to get an appreciation for the computational network biology steps involved in RNA-seq analysis, and how to execute best-practice Cytoscape workflows.

## Objectives:- After this course you should be able to:

- Have more detailed understanding of terms used in network biology.
- Using your RNA-Seq datasets, better visualise enriched terms using Cytoscape.
- Know what tools are available in App Store for RNA-seq, miRNA data and pathway analysis.
- Produce a list of key pathways and differentially expressed genes from an RNA-seq experiment 

## Aims:- During this course you will learn about:

- Cytoscape and specific plugins used for RNA-Seq data analysis. 
- Source for integrating diffrential expressed genes within pathway modules. 
- Hierarchical clustering using String and FunctionalEnrichment apps. 
- Identifying over-represented gene or pathway sets among a list of differentially expressed genes. 
    
## Software installation

You will need to bring an internet-enabled laptop to the course and install the latest versions of Cytoscape before coming to the course

1) Install Cytoscape 

### Windows

* Go to https://apps.cytoscape.org/apps/cluego and click Download or Install (if your cytoscape application is open)
* More on installation, documentation and how to integrate the package with R or Python is described [here:](http://www.ici.upmc.fr/cluego/cluegoDocumentation.shtml)


### Mac

Go to https://apps.cytoscape.org/apps/cluego and click Download or Install (if your cytoscape application is open)

### Linux

Go to https://apps.cytoscape.org/apps/cluego and click Download or Install (if your cytoscape application is open)

2) Install the Cytoscape packages/Apps required for the course

* Go to  [ClueGO](https://apps.cytoscape.org/apps/cluego)
  and  [CluePedia](https://apps.cytoscape.org/apps/cluepedia)
* Click Download or Install (if your cytoscape application is open)

3) Download and unzip this file containing the data for the course

- [Link to Data](https://drive.google.com/file/d/178yZSaG1tTr3akeN0DQ2f1Pu2rMoJkBT/view?usp=sharing)


4) (Optional) Watch these short introductory videos

- [A gentle introduction to Cytoscape](https://www.youtube.com/watch?v=YH-XXHhrv58)
- [Visualisation of Biological Networks using Cytoscape](https://www.youtube.com/watch?v=sqM7ZLhTJoM&list=PLFQS98nmv__zaNXLUQACQu1dHWjtpToKN&index=3)

## Instructors

- Dr. Akshay Bhat, Cancer Bioinformatician
- Dr. Mark Dunning, Bioinformatics Core Director
- Dr. Emily Chambers, Bioinformatics Core Scientist


## Timetable (provisional)

- Session 1 Thursday 12th June 9:00  9:00 - 13:00
  + Importing example cancer RNA-seq counts into Cytoscape
  + Querying pathway and interaction sources using Cytoscape - App. 
  + Styling the network and saving for publication. 
  
- Session 2 Thursday 12th June 13:15 - 15:00
  + Using diffrential expressed gene-lists, perform protein-protein interactions or gene-regulatory networks. 
  + Manipulating and filtering gene-lists, pathways and clustered modules. 
  + Creating sub-networks of specific derived clustered modules. 
  + Enriching gene-list into pathways using FunctionalEnrichment - App. 
  
- Session 3 Thursday 12th June 15:15 - 16:30
  + Pathway clustering using ClueGO - Functional Enrichment
  + Using ClueGO's features for integrating and querying multiple pathway database sources. 
  + Use CluePedia (sister-software for ClueGO) to map gene regulatory information within specific pathways.
  
## Feedback



