# Analysis of next-generation sequencing data in **Cytoscape**

![](sbc.png)

- June 12th University of Sheffield
- Bartolome House, Seminar Room EG03

# Course Overview
The workshop is designed to give you an introduction to using Cytoscape – a network-based software application to perform Bioinformatics pathway enrichment analysis. The particular use-case is that of RNA-seq data, but many of the concepts are used for other types of analysis. There are several reasons why we might want to use Cytoscpae in Bioinformatics:-
•	You can be a non-programmer for analysing your biological datasets generated by “high-throughput sequencing” thorugh the dedicated Cytoscape pacages.
•	The packages within Cytoscape provide a user complete support for different types of “-Omics” datasets. 
 .	e.g. many NGS tools are developed for a Unix environment
•	Some tasks require more memory (RAM) and processor speed than we have on our own machine
•	Some analyses take too long, and whilst they are running everything else on our machine will be slow
•	Eventually you will want to use “High-performance computing” (HPC) - but this requires some background knowledge before you can use it

Creating Network Graphs with Cytoscape

Cytoscape is a tool for viewing and analysing networks (meaning, in this case, any group of entities that are connected in some way). Cytoscape is not too hard to use, but it won't make much sense unless you have a sense of some basic network analysis vocabulary and concepts. A glossary is made below for basic understanding of introduction to network graphs. 
A.	https://github.com/miriamposner/network_analysis_workshop/blob/master/social-network-glossary.md
B.	https://journalofdigitalhumanities.org/1-1/demystifying-networks-by-scott-weingart/

What is this?

These tutorials provide a basic introduction to using Cytoscape to conduct network analysis of biological data. They're intended to be used with the data my students and I collected about early African-American silent film. However, anyone should be able to follow along.

Here's what's included:
•	Creating a basic Cytoscape network (that's this tutorial)
•	Importing a node list and working with node attributes
•	Selecting parts of your network 
•	Publishing your network diagram


## Schedule (Provisonal)

### Monday 30th January, 2023


- 09:30 - 11:00 Introduction and quality assessment
- 11:15 - 12:30 Normalisation and sources of variation
- 12:30 - 13:15 LUNCH
- 13:15 - 15:00 Clustering (UMAP, tSNE etc)
- 15:00 - 16:30 Identifying marker genes and differential expression

## Materials

This course comprises of three sessions. 
- [Session 1 - Introduction to Cytoscape and it's App](session1.Rmd)
- [Part 1 - HTML](session1.nb.html)
- [Session 2 - Functional Enrichment Analysis](session2.Rmd)
- [Part 2 - HTML](session2.nb.html)
- [Session 3 - Enrichment using ClueGO and CluePedia.](session3.Rmd)
- [Part 3 - HTML](session3.nb.html)

## Course coordinator - Dr. Akshay Bhat
## Last modified: 2nd April 2024
